Store creation MVP: add entry point to prologue after account creation

jaclync opened the pull request:
<!-- Remember about a good descriptive title. -->

Part of #7891 
Part of #7879 

⚠️ Please only review this PR after the base PR https://github.com/woocommerce/woocommerce-ios/pull/7906 is approved

<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

In https://github.com/woocommerce/woocommerce-ios/pull/7906, account creation can be initiated from the login prologue (tentatively from the `New to WooCommerce` CTA). This PR integrates account creation completion with store creation, by extracting the store creation webview logic from `StorePickerViewController` to a new `StoreCreationCoordinator`.

### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

- Launch the app
- Log out if needed
- Tap on the `New to WooCommerce?` CTA on the prologue screen --> it should navigate to a form to create a WPCOM account as in the screenshot
- Enter a new email and password, then tap on the submit CTA --> after a bit, it should open the store creation view
- Proceed to finish store creation --> after a new Jetpack site is available, the view should be dismissed and the app should be connected to the new store

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

jaclync committed:
Create `StoreCreationCoordinator` that encapsulates logic for store creation navigation.

jaclync committed:
Remove source-dependent `StorePickerConfiguration` in `SwitchStoreNoticePresenter.presentStoreSwitchedNoticeWhenSiteIsAvailable`.

jaclync committed:
Update store picker entry point to store creation using `StoreCreationCoordinator`.

jaclync committed:
Add store creation entry point to login prologue screen using `StoreCreationCoordinator`.

itsmeichigo commented on WooCommerce/Classes/Authentication/Store Creation/StoreCreationCoordinator.swift:
```diff
@@ -0,0 +1,135 @@
+        case storePicker
+    }
+
+    var navigationController: UINavigationController
```
❓ Is there any case where we want to update this from outside? Should we make it private instead?

jaclync replied:
It's just to conform to `Coordinator` protocol that requires the navigation controller to be mutable:

https://github.com/woocommerce/woocommerce-ios/blob/34c040c9d156ea730c0d10a06c0e03554dfbe5d0/WooCommerce/Classes/Tools/Coordinator.swift#L8

I started an async discussion about this p1666768545000749-slack-CGPNUU63E, in the past I had the same question https://github.com/woocommerce/woocommerce-ios/pull/2669#discussion_r472885799 😄

itsmeichigo commented on WooCommerce/Classes/Authentication/Store Creation/StoreCreationCoordinator.swift:
```diff
@@ -0,0 +1,135 @@
+         analytics: Analytics = ServiceLocator.analytics) {
+        self.source = source
+        self.navigationController = navigationController
+        self.storePickerViewModel = .init(configuration: .standard, //todo
```
❓ Maybe we need the `.login` config here instead?

jaclync replied:
the store picker is shown not only during the login process but also when the app is logged-in when switching stores or launching the app without a connected store. I think `standard` makes more sense here, so that non-Woo sites can be observed for checking if the new site is ready.

alternatively, we can update the `StorePickerViewModel`'s initializer to take in a parameter for whether to support non-Woo sites, since that seems to be the only place where the configuration is used. I just thought this feels a bit out of the scope of the PR.

itsmeichigo commented on WooCommerce/Classes/Authentication/Store Creation/StoreCreationCoordinator.swift:
```diff
@@ -0,0 +1,135 @@
+    func syncSites(forSiteThatMatchesPossibleURLs possibleURLs: Set<String>) async throws -> Site {
+        return try await withCheckedThrowingContinuation { [weak self] continuation in
+            storePickerViewModel.refreshSites(currentlySelectedSiteID: nil) { [weak self] in
+                guard let self else { return }
```
`continuation` should be resumed here to avoid leak:
```suggestion
                guard let self else { return continuation.resume(throwing: <some unknown error> }
```

jaclync replied:
updated in https://github.com/woocommerce/woocommerce-ios/pull/7927/commits/730370922a0fa28289685db85af6e2c461828c26 - I hope we don't have to handle such weak self all the time 😓

itsmeichigo commented on WooCommerce/Classes/Authentication/Store Creation/StoreCreationCoordinator.swift:
```diff
@@ -0,0 +1,135 @@
+            // Shows `My store` tab by default.
+            MainTabBarController.switchToMyStoreTab(animated: true)
+
+            if siteChanged {
+                let presenter = SwitchStoreNoticePresenter(siteID: site.siteID)
+                // The configuration only matters when the switch store notice is shown for `switchingStores` configuration.
+                presenter.presentStoreSwitchedNoticeWhenSiteIsAvailable(configuration: .standard)
+            }
```
❓ Do we really need this? I think we are navigating from store creation so there's no need to handle store switch.

jaclync replied:
Good point, removed in https://github.com/woocommerce/woocommerce-ios/pull/7927/commits/156dd953d936bd423c12027ae75c7c99a645a171 (the site title of a newly created site in the current flow is always empty, which makes the notice look odd anyways)

itsmeichigo reviewed (approved):
This looks good to me 👍 I left a few questions below:

jaclync committed:
Merge branch 'feat/7891-account-creation-ui' into feat/7879-account-creation-to-store-creation

* feat/7891-account-creation-ui:
  Minimize `withCheckedContinuation` logic and add `synchronizeEntities` when handling success.
  Clear email and password error after the corresponding input changes.
  Polish in-progress state of account creation.

jaclync committed:
Show store picker after account creation success instead of showing store creation directly.

jaclync committed:
Remove any chance of switch store notice, the site title for a new site is always empty and looks odd.

jaclync committed:
Merge branch 'trunk' into feat/7879-account-creation-to-store-creation

* trunk:
  Remove local StoreKit configuration
  Adds feature flag
  Check local storage check for already synced products
  Add tests for ProductAction.checkForProducts
  Add tests for loadProductIDs remote request
  Add product action to check if store has any products
  Add remote support for requesting list of product ids
  7908 Add GET request for Just In Time Messages
  7908 Add Swift Concurrency `enqueue` to Remote
  7908 Add Mapper for Just In Time Message Lists
  7908 Add Just In Time Message networking model
  Remove trailing whitespaces from release notes

jaclync committed:
Support different presentation styles of store picker for navigation to store creation.

jaclync committed:
Pass `standard` configuration to `StorePickerViewModel` to include non-Woo sites.

jaclync committed:
StoreCreationCoordinator: throw an error when self is deallocated.

jaclync commented:
@itsmeichigo Other than addressing the CR comments, I've made a few changes to the navigation to accommodate different presentation styles of the store picker. Also, I updated the prologue store creation navigation to show the store picker first before presenting the store creation modally so that the store picker is shown after dismissing the modal (feel free to suggest a better UX). Sorry about the number of changes since the last review 😅

itsmeichigo reviewed (approved):
TIL about store creation from the store picker when switching store - I thought the creation was available only during login 😅 I tested both cases, and the navigation looks good to me 👍 

Probably an issue on the webpage: I tested a case where I picked a domain name and then backed out and tried again - the second time, I wasn't presented with the domain selection page and got stuck in the loading state. Just want to point this out in case this is an issue.


https://user-images.githubusercontent.com/5533851/198176708-7f6f1812-47bc-406a-aa2d-cf2d292d6091.mp4

jaclync commented:
> Probably an issue on the webpage: I tested a case where I picked a domain name and then backed out and tried again - the second time, I wasn't presented with the domain selection page and got stuck in the loading state. Just want to point this out in case this is an issue.

Thanks for testing again and reporting this issue! So far just Sharma and I have seen this p1666328130883459/1666328061.472569-slack-C03L1NF1EA3, I'm glad this issue is further confirmed! It might be related to how the WPCOM plan button opens the next page (where we get stuck) in a new window using `WKWebView`. We plan to make some changes to the web flow for mobile agent, including not opening the domain selection screen in a new window. Hopefully, that fixes this issue.

Merging this PR now!
